cmd/kubeadm/app/phases/kubeconfig/kubeconfig.go

```diff
@@ -77,29 +77,53 @@ func CreateJoinControlPlaneKubeConfigFiles(outDir string, cfg *kubeadmapi.InitCo
 		externaCA = true
 	}
 
-	files := []string{
-		kubeadmconstants.AdminKubeConfigFileName,
-		kubeadmconstants.ControllerManagerKubeConfigFileName,
-		kubeadmconstants.SchedulerKubeConfigFileName,
+	if externaCA {
+		fmt.Printf("[kubeconfig] External CA mode: Using user provided %s\n", kubeadmconstants.AdminKubeConfigFileName)
+		fmt.Printf("[kubeconfig] External CA mode: Using user provided %s\n", kubeadmconstants.ControllerManagerKubeConfigFileName)
+		fmt.Printf("[kubeconfig] External CA mode: Using user provided %s\n", kubeadmconstants.SchedulerKubeConfigFileName)
+		return nil
 	}
 
-	for _, file := range files {
-		if externaCA {
-			fmt.Printf("[kubeconfig] External CA mode: Using user provided %s\n", file)
-			continue
-		}
-		if err := createKubeConfigFiles(outDir, cfg, file); err != nil {
-			return err
-		}
+	if err := createKubeConfigFiles(outDir, cfg, kubeadmconstants.AdminKubeConfigFileName); err != nil {
+		return err
+	}
+
+	if err := createKubeConfigFiles(outDir, cfg, kubeadmconstants.ControllerManagerKubeConfigFileName); err != nil {
+		return err
+	}
+
+	file := filepath.Join(outDir, kubeadmconstants.ControllerManagerKubeConfigFileName)
+	if err := os.Chown(file, int(cfg.ControllerManager.UID), int(cfg.ControllerManager.GID)); err != nil {
+		return err
+	}
+
+	if err := createKubeConfigFiles(outDir, cfg, kubeadmconstants.SchedulerKubeConfigFileName); err != nil {
+		return err
 	}
+	fmt.Printf("[kubeconfig] setting owner of kubeconfig file %s to uid: %d gid: %d\n", kubeadmconstants.AdminKubeConfigFileName, cfg.Scheduler.UID, cfg.Scheduler.GID)
+	file = filepath.Join(outDir, kubeadmconstants.SchedulerKubeConfigFileName)
+	if err := os.Chown(file, int(cfg.Scheduler.UID), int(cfg.Scheduler.GID)); err != nil {
```

Comment:
chown-ing a kubeconfig file is not in scope of https://github.com/kubernetes/kubeadm/issues/1367

Reply:
This is only the file we just created?
What do you propose instead?

Reply:
around 2018 there was a discussion whether kubeadm should chown the files it creates, and everybody voted no.
is there are related issue to this PR?

Reply:
there's no mention of chown in the linked issue, can you provide some more context for the 2018 discussion?
I would expect the components to own their own config.

Reply:
> there's no mention of chown in the linked issue, can you provide some more context for the 2018 discussion?

i had a PR at some point to make the kubeadm admin.conf be chowed for a target UID / GID which helps the last manual step after `kubeadm init` finishes and kubeconfigs must be copied to `~/.kube`.
the idea was dropped with the argument that users can do whatever they want with kubeconfigs and kubeadm should not be in charge of chowing or moving them.

notably chown doesn't work for Windows, so the current code in this PR is not portable.

> I would expect the components to own their own config.

this writes a kubeconfig on disk and chowns it for a certain UID / GID, which is strange and i'm confused why its needed. on a kubeadm host /etc/kubernetes/* is owned by root. https://github.com/kubernetes/kubeadm/issues/1367 is about moving components running in containers away from root.

Reply:
If a component is running as non root (like kubescheduler in this case) and mounts hostVols (like kubeconfig file in this case) it needs permissions to access those files. So when we run kube-scheduler as uid 2000 the scheduler.conf file must be accessible to user 2000, hence the chown is required.

Reply:
> notably chown doesn't work for Windows, so the current code in this PR is not portable.

none of the control plane components support running on windows to begin with?

only kubelet, containerd or docker, and possibly some of kube-proxy.

Reply:
also: `hostPath` volumes do not support altering permissions intentionally, so you must chown them from something else with permission (such as `kubeadm init`), I don't think you can move the pod to rootless without doing this.

Reply:
commenting in the main thread.